doc/manual/metaprogramming.rst

```diff
@@ -226,89 +322,112 @@ the important distinction between the way ``a`` and ``b`` are used:
-      @eval ($op)(a,b,c) = ($op)(($op)(a,b),c)
-    end
+Macros
+------
 
-The :obj:`@eval` macro rewrites this call to be precisely equivalent to the
-above longer versions. For longer blocks of generated code, the
-expression argument given to :obj:`@eval` can be a block::
+Macros are the analogue of functions for expression generation at
+compile time. Just as a function maps a tuple of argument values to a
+return value, a macro maps a tuple of argument *expressions* to a returned
+*expression*.
 
-    @eval begin
-      # multiple lines
-    end
+Why macros
+~~~~~~~~~~
 
-Interpolating into an unquoted expression is not supported and will
-cause a compile-time error:
+Here is an extraordinarily simple macro:
 
 .. doctest::
 
-    julia> $a + b
-    ERROR: unsupported or misplaced expression $
+    julia> macro sayhello(name)
+	       :( println("Hello, ", $name, "!") )
+	   end
 
-.. _man-macros:
+The "return value" of this macro is the *interpolated* expression; that is,
+with the value of the variable ``name`` passed directly as an argument to
+:func:`println`. We can verify this with the **extremely
+important** function :func:`macroexpand`:
 
-Macros
-------
+.. doctest::
 
-Macros are the analogue of functions for expression generation at
-compile time. Just as functions map a tuple of argument values to a 
-return value, macros map a tuple of argument *expressions* to a returned
-*expression*. They allow the programmer to arbitrarily transform the
-written code to a resulting expression, which then takes the place of
-the macro call in the final syntax tree. Macros are invoked with the
-following general syntax::
+    julia> ex = macroexpand( :(@sayhello("human")) )
+    :(println("Hello, ","human","!"))
+
+    julia> typeof(ex)
+    Expr
+
+:func:`macroexpand` takes a *quoted* macro and returns the resulting
+:obj:`Expr`\ession *after interpolation*.
+
+**Hold up.**
+
+We have already seen a function ``f(::Expr...) -> Expr`` in a previous
+section. In fact, :func:`macroexpand` is such a function itself;
+evaluating the result of :func:`macroexpand` will give the expected
+output:
+
+.. doctest::
+
+    julia> eval(macroexpand( quote @sayhello("human") end ))
+    Hello, human!
+
+So, why are macros even necessary? As a simplifying analogy, let's
+define ``The Full Julia Compiler`` as ``eval, but much faster``.
+We actually want to *avoid* calling the :func:`eval` directly in
+"real" code, because it will be exceptionally slow compared to using
+``The Full Julia Compiler``. But, if we are generating code, how
+how can execute our expression without calling :func:`eval`?
+
+The answer is macros.
+
+Macros are useful because they allow the programmer to
+transform generated code into expanded expressions
+*before* the full program is compiled and executed.
+
+Macro invocation
+~~~~~~~~~~~~~~~~
+
+Macros are invoked with the following general syntax::
 
     @name expr1 expr2 ...
     @name(expr1, expr2, ...)
```

Comment:
It's actually that `f (x)` (with a space between `f` and the parenthesis) is going to be disallowed, because of the issue described in the next paragraph.

I know what you mean about the length of the page, but since this is a "gotcha" I tend to think it needs to be described.